File: backend/app/schemas/call_back_dashboard.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

CallBackBucket = Literal["overdue", "due_today", "upcoming", "no_call_date"]
CallBackStatus = Literal["overdue", "due_today", "upcoming", "no_call_date", "completed"]


class CallBackDashboardQuoteRow(BaseModel):
    id: int
    quote_ref: str | None
    eworks_quote_id: int
    customer_name: str | None
    site_address: str | None
    quote_value: float | None
    status: str | None
    status_name: str | None
    tags: list[str] = Field(default_factory=list)
    created_on: str | None
    last_updated_on: str | None
    days_since_updated: int
    assigned_name: str | None
    assigned_email: str | None
    call_note: str | None
    last_called_at: str | None
    next_call_at: str | None
    call_status: CallBackStatus
    quote_detail_link: str


class CallBackDashboardCategory(BaseModel):
    count: int
    value: float
    quotes: list[CallBackDashboardQuoteRow]


class CallBackDashboardCategories(BaseModel):
    overdue: CallBackDashboardCategory
    due_today: CallBackDashboardCategory
    upcoming: CallBackDashboardCategory
    no_call_date: CallBackDashboardCategory


class CallBackDashboardTotals(BaseModel):
    call_back_quotes: int
    total_quote_value: float
    overdue_calls: int
    due_today_calls: int
    upcoming_calls: int
    no_call_date: int
    average_age_days: float


class CallBackDashboardRead(BaseModel):
    totals: CallBackDashboardTotals
    categories: CallBackDashboardCategories


class CallBackTrackingPatch(BaseModel):
    assigned_user_id: str | None = None
    assigned_name: str | None = None
    assigned_email: str | None = None
    call_note: str | None = None
    last_called_at: str | None = None
    next_call_at: str | None = None


class CallBackTrackingRead(BaseModel):
    id: str
    synced_quote_id: int | None
    eworks_quote_id: int
    quote_ref: str | None
    assigned_user_id: str | None
    assigned_name: str | None
    assigned_email: str | None
    call_note: str | None
    last_called_at: str | None
    next_call_at: str | None
    call_status: str | None

File: frontend/src/lib/call-back-dashboard.ts
import { apiFetch } from "@/lib/api";

export type CallBackBucket = "overdue" | "due_today" | "upcoming" | "no_call_date";
export type CallBackStatus = "overdue" | "due_today" | "upcoming" | "no_call_date" | "completed";

export type CallBackDashboardQuote = {
  id: number;
  quote_ref: string | null;
  eworks_quote_id: number;
  customer_name: string | null;
  site_address: string | null;
  quote_value: number | null;
  status: string | null;
  status_name: string | null;
  tags: string[];
  created_on: string | null;
  last_updated_on: string | null;
  days_since_updated: number;
  assigned_name: string | null;
  assigned_email: string | null;
  call_note: string | null;
  last_called_at: string | null;
  next_call_at: string | null;
  call_status: CallBackStatus;
  quote_detail_link: string;
};

export type CallBackDashboardCategory = {
  count: number;
  value: number;
  quotes: CallBackDashboardQuote[];
};

export type CallBackDashboardTotals = {
  call_back_quotes: number;
  total_quote_value: number;
  overdue_calls: number;
  due_today_calls: number;
  upcoming_calls: number;
  no_call_date: number;
  average_age_days: number;
};

export type CallBackDashboard = {
  totals: CallBackDashboardTotals;
  categories: Record<CallBackBucket, CallBackDashboardCategory>;
};

export type CallBackTrackingPatch = {
  assigned_user_id?: string | null;
  assigned_name?: string | null;
  assigned_email?: string | null;
  call_note?: string | null;
  last_called_at?: string | null;
  next_call_at?: string | null;
};

export async function getCallBackDashboard(
  apiBase: "manager" | "admin",
  search?: string,
): Promise<CallBackDashboard> {
  const params = new URLSearchParams();
  if (search?.trim()) params.set("search", search.trim());
  const qs = params.toString();
  const res = await apiFetch<CallBackDashboard>(
    `/api/v1/${apiBase}/call-back-dashboard${qs ? `?${qs}` : ""}`,
  );
  const data = res.data;
  if (!data?.totals || !data?.categories) {
    throw new Error("Invalid Call Back dashboard response");
  }
  return data;
}

export async function patchCallBackTracking(quoteId: number, patch: CallBackTrackingPatch): Promise<void> {
  await apiFetch(`/api/v1/call-back-quotes/${quoteId}/tracking`, {
    method: "PATCH",
    body: JSON.stringify(patch),
  });
}

export const BUCKET_LABELS: Record<CallBackBucket, string> = {
  overdue: "Overdue",
  due_today: "Due Today",
  upcoming: "Upcoming",
  no_call_date: "No Call Date",
};

export const CALL_STATUS_LABELS: Record<CallBackStatus, string> = {
  overdue: "Overdue",
  due_today: "Due today",
  upcoming: "Upcoming",
  no_call_date: "No call date",
  completed: "Completed",
};
